File: app/src/main/java/com/ghostnexora/vpn/tunnel/PhysicalNetworkSocketConnector.java
package com.ghostnexora.vpn.tunnel;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;

import java.io.IOException;
import java.net.Inet4Address;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;

import kotlin.Unit;
import kotlin.jvm.functions.Function1;

/**
 * Creates all Java/JSch transport sockets outside the Android VPN TUN.
 *
 * Protection order is intentionally strict:
 * 1) VpnService.protect(Socket)
 * 2) bind the socket to a NET_CAPABILITY_NOT_VPN Network when available
 * 3) connect only after both routing protections have been applied
 *
 * The owning application UID is also excluded by GhostVpnService.Builder,
 * providing an independent third loop-prevention layer.
 */
public final class PhysicalNetworkSocketConnector {
    private static final int DEFAULT_CONNECT_TIMEOUT_MS = 20_000;
    private static final int PER_ADDRESS_TIMEOUT_MS = 8_000;

    private final ConnectivityManager connectivityManager;
    private final Function1<? super String, Unit> onStatus;
    private final Set<Network> knownPhysicalNetworks = ConcurrentHashMap.newKeySet();
    private final AtomicBoolean protectionReported = new AtomicBoolean(false);

    private final ConnectivityManager.NetworkCallback physicalNetworkCallback =
            new ConnectivityManager.NetworkCallback() {
                @Override
                public void onAvailable(Network network) {
                    knownPhysicalNetworks.add(network);
                }

                @Override
                public void onCapabilitiesChanged(Network network, NetworkCapabilities capabilities) {
                    boolean physicalInternet =
                            capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                                    && capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN);
                    if (physicalInternet) {
                        knownPhysicalNetworks.add(network);
                    } else {
                        knownPhysicalNetworks.remove(network);
                    }
                }

                @Override
                public void onLost(Network network) {
                    knownPhysicalNetworks.remove(network);
                }
            };

    public PhysicalNetworkSocketConnector(Context context, Function1<? super String, Unit> onStatus) {
        Context appContext = context != null ? context.getApplicationContext() : null;
        this.connectivityManager = appContext != null
                ? (ConnectivityManager) appContext.getSystemService(Context.CONNECTIVITY_SERVICE)
                : null;
        this.onStatus = onStatus != null ? onStatus : ignored -> Unit.INSTANCE;
        registerPhysicalNetworkCallback();
    }

    public PhysicalNetworkSocketConnector(Context context) {
        this(context, ignored -> Unit.INSTANCE);
    }

    public Socket connect(String host, int port) throws IOException {
        return connect(host, port, DEFAULT_CONNECT_TIMEOUT_MS);
    }

    public Socket connect(String host, int port, int timeoutMs) throws IOException {
        if (host == null || host.trim().isEmpty()) {
            throw new IllegalArgumentException("El host de transporte no puede estar vacío");
        }
        if (port < 1 || port > 65_535) {
            throw new IllegalArgumentException("El puerto de transporte es inválido");
        }

        long deadlineNanos = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeoutMs);
        List<Throwable> failures = new ArrayList<>();
        List<Network> networks = physicalNetworks();

        for (int networkIndex = 0; networkIndex < networks.size(); networkIndex++) {
            Network network = networks.get(networkIndex);
            List<InetAddress> addresses = resolve(network, host, failures);
            if (addresses.isEmpty()) {
                continue;
            }

            status("[NETWORK] DNS físico · " + host + " → " + joinAddresses(addresses));
            for (int addressIndex = 0; addressIndex < addresses.size(); addressIndex++) {
                InetAddress address = addresses.get(addressIndex);
                int remainingMs = remainingTimeoutMs(deadlineNanos);
                if (remainingMs <= 0) {
                    break;
                }

                Socket socket = configuredAndProtectedSocket();
                try {
                    network.bindSocket(socket);
                    status("[NETWORK] Intento TCP físico " + (networkIndex + 1) + "." +
                            (addressIndex + 1) + " · " + address.getHostAddress() + ":" + port +
                            " · " + networkLabel(network));
                    long startedAt = System.nanoTime();
                    socket.connect(new InetSocketAddress(address, port), remainingMs);
                    long latencyMs = Math.max(1L,
                            TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startedAt));
                    status("[NETWORK] Socket TCP físico conectado · " + address.getHostAddress() +
                            ":" + port + " · " + latencyMs + " ms · " + networkLabel(network));
                    return socket;
                } catch (Throwable error) {
                    failures.add(error);
                    closeQuietly(socket);
                    status("[NETWORK] IP no disponible · " + address.getHostAddress() + ":" + port +
                            " · " + shortMessage(error));
                }
            }
        }

        // Vendor fallback: still fail-closed because protect(Socket) is mandatory
        // even when Android does not expose a bindable NOT_VPN Network instance.
        List<InetAddress> fallbackAddresses = resolve(null, host, failures);
        for (int index = 0; index < fallbackAddresses.size(); index++) {
            InetAddress address = fallbackAddresses.get(index);
            int remainingMs = remainingTimeoutMs(deadlineNanos);
            if (remainingMs <= 0) {
                break;
            }

            Socket socket = configuredAndProtectedSocket();
            try {
                status("[NETWORK] Intento TCP protegido " + (index + 1) + " · " +
                        address.getHostAddress() + ":" + port);
                socket.connect(new InetSocketAddress(address, port), remainingMs);
                status("[NETWORK] Socket TCP conectado con protect(Socket) · " +
                        address.getHostAddress() + ":" + port);
                return socket;
            } catch (Throwable error) {
                failures.add(error);
                closeQuietly(socket);
            }
        }

        Throwable last = failures.isEmpty() ? null : failures.get(failures.size() - 1);
        int attempted = Math.max(1, failures.size());
        throw new IOException(
                "[TCP-ALL-FAILED] No fue posible conectar con ninguna IP de " + host + ":" + port +
                        " tras " + attempted + " intento(s).",
                last
        );
    }

    private void registerPhysicalNetworkCallback() {
        if (connectivityManager == null) {
            return;
        }
        NetworkRequest request = new NetworkRequest.Builder()
                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
                .build();
        try {
            connectivityManager.registerNetworkCallback(request, physicalNetworkCallback);
        } catch (Throwable error) {
            status("[NETWORK] Registro de redes físicas no disponible · " + shortMessage(error));
        }
    }

    private List<Network> physicalNetworks() {
        if (connectivityManager == null) {
            return new ArrayList<>();
        }
        List<Network> result = new ArrayList<>();
        Network active = connectivityManager.getActiveNetwork();
        if (active != null && isPhysicalInternetNetwork(active)) {
            result.add(active);
        }
        for (Network network : knownPhysicalNetworks) {
            if (!network.equals(active) && isPhysicalInternetNetwork(network)) {
                result.add(network);
            }
        }
        return result;
    }

    private List<InetAddress> resolve(Network network, String host, List<Throwable> failures) {
        try {
            InetAddress[] raw = network != null ? network.getAllByName(host) : InetAddress.getAllByName(host);
            Map<String, InetAddress> unique = new LinkedHashMap<>();
            for (InetAddress address : raw) {
                unique.put(address.getHostAddress(), address);
            }
            List<InetAddress> result = new ArrayList<>(unique.values());
            result.sort(Comparator.comparingInt(address -> address instanceof Inet4Address ? 0 : 1));
            return result;
        } catch (Throwable error) {
            failures.add(error);
            return new ArrayList<>();
        }
    }

    private int remainingTimeoutMs(long deadlineNanos) {
        long remaining = TimeUnit.NANOSECONDS.toMillis(deadlineNanos - System.nanoTime());
        return (int) Math.min(remaining, PER_ADDRESS_TIMEOUT_MS);
    }

    private boolean isPhysicalInternetNetwork(Network network) {
        if (connectivityManager == null) {
            return false;
        }
        NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(network);
        return capabilities != null
                && capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                && capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN);
    }

    private String networkLabel(Network network) {
        NetworkCapabilities capabilities = connectivityManager != null
                ? connectivityManager.getNetworkCapabilities(network)
                : null;
        if (capabilities != null && capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
            return "datos móviles";
        }
        if (capabilities != null && capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
            return "Wi-Fi";
        }
        if (capabilities != null && capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
            return "Ethernet";
        }
        return "red física";
    }

    private Socket configuredAndProtectedSocket() throws IOException {
        Socket socket = configuredSocket();
        if (!OutboundSocketProtection.protect(socket)) {
            closeQuietly(socket);
            throw new IOException(
                    "[VPN-LOOP-001] Android rechazó VpnService.protect(Socket); " +
                            "el transporte se detuvo para evitar un bucle hacia el TUN"
            );
        }
        if (protectionReported.compareAndSet(false, true)) {
            status("[NETWORK] Sockets de transporte excluidos del TUN con VpnService.protect(Socket)");
        }
        return socket;
    }

    private static Socket configuredSocket() throws IOException {
        Socket socket = new Socket();
        socket.setTcpNoDelay(true);
        socket.setKeepAlive(true);
        socket.setReuseAddress(true);
        return socket;
    }

    private void status(String message) {
        onStatus.invoke(message);
    }

    private static String joinAddresses(List<InetAddress> addresses) {
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < addresses.size(); i++) {
            if (i > 0) {
                builder.append(", ");
            }
            String value = addresses.get(i).getHostAddress();
            builder.append(value != null ? value : "");
        }
        return builder.toString();
    }

    private static void closeQuietly(Socket socket) {
        try {
            socket.close();
        } catch (Throwable ignored) {
        }
    }

    private static String shortMessage(Throwable error) {
        Throwable current = error;
        while (current != null) {
            String message = current.getMessage();
            if (message != null && !message.trim().isEmpty()) {
                String normalized = message.replace('\n', ' ').trim();
                return normalized.length() <= 220 ? normalized : normalized.substring(0, 220);
            }
            current = current.getCause();
        }
        return error.getClass().getSimpleName();
    }
}

File: app/src/main/java/com/ghostnexora/vpn/tunnel/PhysicalNetworkSocketConnector.kt
package com.ghostnexora.vpn.tunnel

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import java.io.IOException
import java.net.Inet4Address
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.Socket
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.TimeUnit

/**
 * Opens transport sockets on a real Android network instead of letting the
 * process default decide after the VPN TUN is active.
 *
 * The application UID is already excluded from the TUN by GhostVpnService.
 * Binding every SSH/proxy socket to a NOT_VPN network adds a second routing
 * guarantee and also lets DNS resolution use that same physical network.
 */
internal class PhysicalNetworkSocketConnector(
    context: Context?,
    private val onStatus: (String) -> Unit = {}
) {
    private val connectivityManager = context
        ?.applicationContext
        ?.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
    private val knownPhysicalNetworks = ConcurrentHashMap.newKeySet<Network>()

    private val physicalNetworkCallback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) {
            knownPhysicalNetworks += network
        }

        override fun onCapabilitiesChanged(
            network: Network,
            capabilities: NetworkCapabilities
        ) {
            val physicalInternet =
                capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
                    capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
            if (physicalInternet) {
                knownPhysicalNetworks += network
            } else {
                knownPhysicalNetworks -= network
            }
        }

        override fun onLost(network: Network) {
            knownPhysicalNetworks -= network
        }
    }

    init {
        connectivityManager?.let { manager ->
            val request = NetworkRequest.Builder()
                .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
                .addCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
                .build()
            runCatching {
                manager.registerNetworkCallback(request, physicalNetworkCallback)
            }.onFailure { error ->
                onStatus(
                    "[NETWORK] Registro de redes físicas no disponible · ${error.shortMessage()}"
                )
            }
        }
    }

    fun connect(host: String, port: Int, timeoutMs: Int = DEFAULT_CONNECT_TIMEOUT_MS): Socket {
        require(host.isNotBlank()) { "El host de transporte no puede estar vacío" }
        require(port in 1..65535) { "El puerto de transporte es inválido" }

        val deadlineNanos = System.nanoTime() + TimeUnit.MILLISECONDS.toNanos(timeoutMs.toLong())
        val failures = mutableListOf<Throwable>()
        val networks = physicalNetworks()

        for ((networkIndex, network) in networks.withIndex()) {
            val addresses = resolve(network, host, failures)
            if (addresses.isEmpty()) continue

            onStatus(
                "[NETWORK] DNS físico · $host → ${addresses.joinToString { it.hostAddress.orEmpty() }}"
            )
            for ((addressIndex, address) in addresses.withIndex()) {
                val remainingMs = remainingTimeoutMs(deadlineNanos)
                if (remainingMs <= 0) break

                val socket = configuredSocket()
                try {
                    network.bindSocket(socket)
                    onStatus(
                        "[NETWORK] Intento TCP físico ${networkIndex + 1}.${addressIndex + 1} · " +
                            "${address.hostAddress}:$port · ${networkLabel(network)}"
                    )
                    val startedAt = System.nanoTime()
                    socket.connect(InetSocketAddress(address, port), remainingMs)
                    val latencyMs = TimeUnit.NANOSECONDS
                        .toMillis(System.nanoTime() - startedAt)
                        .coerceAtLeast(1L)
                    onStatus(
                        "[NETWORK] Socket TCP físico conectado · ${address.hostAddress}:$port · " +
                            "$latencyMs ms · ${networkLabel(network)}"
                    )
                    return socket
                } catch (error: Throwable) {
                    failures += error
                    runCatching { socket.close() }
                    onStatus(
                        "[NETWORK] IP no disponible · ${address.hostAddress}:$port · " +
                            error.shortMessage()
                    )
                }
            }
        }

        // Defensive fallback for devices whose vendor connectivity service does
        // not expose a NOT_VPN Network object. The app UID remains disallowed
        // from the VPN, so this socket still cannot recurse through the TUN.
        val fallbackAddresses = resolve(network = null, host = host, failures = failures)
        for ((index, address) in fallbackAddresses.withIndex()) {
            val remainingMs = remainingTimeoutMs(deadlineNanos)
            if (remainingMs <= 0) break

            val socket = configuredSocket()
            try {
                onStatus(
                    "[NETWORK] Intento TCP con bypass de aplicación ${index + 1} · " +
                        "${address.hostAddress}:$port"
                )
                socket.connect(InetSocketAddress(address, port), remainingMs)
                onStatus("[NETWORK] Socket TCP conectado por bypass propio · ${address.hostAddress}:$port")
                return socket
            } catch (error: Throwable) {
                failures += error
                runCatching { socket.close() }
            }
        }

        val last = failures.lastOrNull()
        val attempted = failures.size.coerceAtLeast(1)
        throw IOException(
            "[TCP-ALL-FAILED] No fue posible conectar con ninguna IP de $host:$port " +
                "tras $attempted intento(s).",
            last
        )
    }

    private fun physicalNetworks(): List<Network> {
        val manager = connectivityManager ?: return emptyList()
        val active = manager.activeNetwork
        return buildList {
            if (active != null && manager.isPhysicalInternetNetwork(active)) add(active)
            knownPhysicalNetworks.forEach { network ->
                if (network != active && manager.isPhysicalInternetNetwork(network)) add(network)
            }
        }
    }

    private fun resolve(
        network: Network?,
        host: String,
        failures: MutableList<Throwable>
    ): List<InetAddress> = runCatching {
        val resolved = if (network != null) network.getAllByName(host) else InetAddress.getAllByName(host)
        resolved
            .distinctBy { it.hostAddress }
            .sortedBy { if (it is Inet4Address) 0 else 1 }
    }.onFailure { failures += it }.getOrDefault(emptyList())

    private fun remainingTimeoutMs(deadlineNanos: Long): Int = TimeUnit.NANOSECONDS
        .toMillis(deadlineNanos - System.nanoTime())
        .coerceAtMost(PER_ADDRESS_TIMEOUT_MS.toLong())
        .toInt()

    private fun ConnectivityManager.isPhysicalInternetNetwork(network: Network): Boolean {
        val capabilities = getNetworkCapabilities(network) ?: return false
        return capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
            capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_VPN)
    }

    private fun networkLabel(network: Network): String {
        val capabilities = connectivityManager?.getNetworkCapabilities(network)
        return when {
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) == true -> "datos móviles"
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) == true -> "Wi-Fi"
            capabilities?.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) == true -> "Ethernet"
            else -> "red física"
        }
    }

    private fun configuredSocket(): Socket = Socket().apply {
        tcpNoDelay = true
        keepAlive = true
        reuseAddress = true
    }

    private fun Throwable.shortMessage(): String = generateSequence(this) { it.cause }
        .mapNotNull { it.message?.takeIf(String::isNotBlank) }
        .firstOrNull()
        .orEmpty()
        .replace('\n', ' ')
        .take(220)
        .ifBlank { javaClass.simpleName }

    private companion object {
        const val DEFAULT_CONNECT_TIMEOUT_MS = 20_000
        const val PER_ADDRESS_TIMEOUT_MS = 8_000
    }
}
